accepted/2023/net8.0-browser-tfm.md

```diff
@@ -0,0 +1,187 @@
+
+One of the reasons we were hesitant to ask for a TFM in the initial release of Blazor was that we were concerned about people writing .NET webassembly specific code that did not work with Blazor Server. Over time, we have determined that it is not a real risk for two reasons:
+* Blazor Server has a healthy user base, which makes it desirable for library authors to target net8.0.
+* A core piece of Blazor functionality (prerendering) depends on being able to run the application on net8.0.
+
+Given this, we consider unlikely that developers will author wasm specific code (net8.0-wasm) without also including a net8.0 version, as it will make it incompatible with any other Blazor flavor.
+
+In addition this, we also have the experience with Blazor Hybrid and Maui where developers already have to work with multiple TFMs (one per platform) and Blazor Hybrid being part of that ecosystem needs to deal with that too.
+
+Finally, we are looking at two new core experiences for which a TFM is heavily desirable:
+* Blazor Web (United): A version of Blazor that combines static rendering + progressive enhancement to deliver the best of both worlds, webassembly and server. In this model, we have an application in a single project that starts as a server application, and can then transition to run directly in the browser via webassembly.
+  * In this model customers can author components that contain webassembly specific dependencies and implementations without requiring a common contract with the server (shared abstractions in a separate library) and can have different sets of dependencies without requiring a separate package/dll.
+    * For example, the server specific code can reference primitives in ASP.NET Core without that dependency flowing to webassembly.
+  * This also applies to library authors that create Razor Class Libraries, as they can provide an implementation for Blazor Server and Webassembly in a single nuget package.
+  * For more context click to see a demo [![here](https://img.youtube.com/vi/48G_CEGXZZM/3.jpg)](https://www.youtube.com/watch?v=48G_CEGXZZM)
+* Blazor Universal: A single project type that produces both MAUI and web applications. Multitargeting across TFMs is how we can produce multiple outputs for these different platforms.
+
+## Scenarios and User Experience
+
+* Projects can have conventions based on the TFM (like MAUI)
+  ![image](https://user-images.githubusercontent.com/6995051/219371569-0699ae29-8bbf-4e93-b1e1-904dc6e3616d.png)
+
+* Projects can have specific dependencies for the browser (like MAUI)
+  ```xml
+  <ItemGroup Condition="'$(Browser)' == 'true'">
+    <Reference Include="Microsoft.AspNetCore.Components.WebAssembly" />
+    <Reference Include="System.Net.Http.Json" />
+  </ItemGroup>
```

Comment:
Does this syntax actually work? I thought we can't do that due to MSBuild evaluation order. /cc @dsplaisted

You can vary dependencies on the platform, but I think you'll need to use the MSBuild intrinsic functions for parsing the platform name out of the TFM.

Reply:
This will likely work, though I don't know how the `Browser` property is set.  It works because item evaluation happens in a later phase than property evaluation.  So if this was properties in a `PropertyGroup` it wouldn't work.

So yes it works in this case, but it's also confusing and if people see this pattern they will probably try to apply it to properties where it won't work.

Reply:
Thanks for bringing this up.

If there are concerns about the correctness or reliability of this technique it would be great to dig into that, since we took this approach in our prototype (where it appears to work perfectly) and are taking a reliance on it as a core part of how these projects will be structured. I *thought* this was equivalent to what MAUI does but if this is mistaken we should try to clarify.

Is there a different/better way to vary things based on TFM?

Reply:
I read up a bit on this last night, the thing that always works is the TFM if I am not mistaken, so we won't use the `$(Browser) == 'true'` pattern but the `$(TargetFramework) == 'net8.0-browser'` pattern, and that should work.

Reply:
The problem with `$(TargetFramework) == 'net8.0-browser'` is that it stops working when you retarget to .NET 9 (`net9.0-browser`).

The "right" way to do it is `'$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'browser'`, but that's quite ugly to put in a project file.

So there's no great solution.  If the main scenario you see for this is to modify items and not properties, it might be OK to go with `'$(Browser)' == 'true'` or `'$(TargetPlatformIdentifier)' == 'browser'`.

Reply:
@dsplaisted fair enough, I don't remember the exact details of how we do it in Maui, but in any case, I believe we can have the "cleaner" pattern for dependencies if we want to.

Maui does the "ugly" suggestion in the template, for reference the Blazor Maui template (although it's the same on the maui one)
https://github.com/dotnet/maui/blob/main/src/Templates/src/templates/maui-blazor/MauiApp.1.csproj#L25-L30

Reply:
I suspect people in their apps, very likely use the target framework directly and do a round of find and replace when they update.

Reply:
Keep in mind that some people may retarget using the Visual Studio project properties UI, and not realize that they need to  open the project file and update other places manually.

Reply:
> Keep in mind that some people may retarget using the Visual Studio project properties UI, and not realize that they need to open the project file and update other places manually.

That's fair, we'll have to check how Maui deals with these types of things, we might follow the same approach and use the MSBuild expression or provide a helper property. We'll have to make a call when we figure out the details for the template.

@dsplaisted won't the property work against other properties if we define it earlier in the props file for the SDK? AFAIK, properties still get evaluated in the order in which they appear in the file, isn't it? I know there are like 6 stages of evaluation, so I understand that it might be more nuanced than what I suggested.

Reply:
>  it might be OK to go with '$(Browser)' == 'true' or '$(TargetPlatformIdentifier)' == 'browser'.

I thought that ` '$(TargetPlatformIdentifier)' == 'browser' ` should already work today, as the .NET SDK already populates that property: https://github.com/dotnet/sdk/blob/9f43a4a55078a22daac527c99e98d26e827a0ee2/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets#L62

However, I see that it's in a targets file. But I'm 99% sure that ` '$(TargetFrameworkIdentifier)' == '.NETFramework' ` works in a csproj, as I use it when I need to multi-target netfx. Maybe there's some inner-build magic going on.

Anyway, I just wanted to point out that I think it will work already (in a csproj), I just don't understand how.

Reply:
Comparing against the `$(TargetFrameworkIdentifier)` works for item / ItemGroup conditions.  It doesn't work for properties / PropertyGroups.

> won't the property work against other properties if we define it earlier in the props file for the SDK? AFAIK, properties still get evaluated in the order in which they appear in the file, isn't it? I know there are like 6 stages of evaluation, so I understand that it might be more nuanced than what I suggested.

If the property depends on the `TargetFramework`, then you can't set it in a .props file, since the `TargetFramewrok` won't be set yet.

Reply:
Given that we wouldn't expect the browser tfm to have versions we can be a bit more creative.

```xml
<TargetFramework>browser</TargetFramework>
<TargetFrameworkVersion>8.0</TargetFrameworkVersion>
```

Then the SDK logic that can "special case" `browser` and not populate the TargetFrameworkVersion, but use whatever is provided. 

Then you'd be able to use `'(TargetFramework)' == 'browser'` like constructs and it'll always work. 

The disadvantage is that this cannot be extended to a general solution.

Reply:
@dsplaisted have we considered using a naming convention for `.props` and `.targets` that are automatically imported, like:

* `NETCoreApp.props`
* `NETFramework.props`
* `macOS.props`
* `ios.props`
* `windows.props`
* `windows10.0`.props`
* `NETCoreApp-windows.props`
...

Like naively, this could work in this exists somewhere in the SDK:

```xml
<Project>
    <PropertyGroup>
        <FrameworkProps>$([MSBuild]::GetTargetFrameworkIdentifier($(TargetFramework))).props</FrameworkProps>
        <PlatformProps>$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework))).props</PlatformProps>
    </PropertyGroup>

    <Import Project="$(FrameworkProps)" Condition="Exists('$(FrameworkProps)')" />
    <Import Project="$(PlatformProps)" Condition="Exists('$(PlatformProps)')" />
</Project>
```

Reply:
We discussed this today and added a note to the document.